### ISSUE TITLE
BitTorrent Announce Request FireWall Alerts With VPN

### ISSUE THREAD

**Issue by ttikaman** [binhex/arch-qbittorrentvpn#80]
Hello,

I have this container configured using Mullvad VPN which I can see being added correctly in the logs.
`2021-02-09 14:13:52 net_route_v4_add: <<VPN_IP>>/32 via 172.17.0.1 dev [NULL] table 0 metric -1`

From within the container I can also confirm my requests are coming from a VPN exit node.
```
sh-5.1# curl icanhazip.com
66.115.189.144
```

However, I am receiving firewall alerts titled "GPL P2P BitTorrent announce request" on regular intervals, example alert below. Is this the expected behavior or should all traffic be routed through the VPN?

Thanks!

<img width="1291" alt="Example" src="https://user-images.githubusercontent.com/78826868/107436824-10bc2280-6afc-11eb-8c7a-097f4f0caf86.png">

**Comment by binhex** [comment]
no its not expected but it is possible, this is because the port for the web ui runs on port 8080 and thus must be accessible from any ip address in order to connect to it outside of your lan (reverse proxy, port forward etc). if your tracker just happens to also be running on port 8080 then qbittorrent can talk to the tracker over your lan side connection.

sadly there is not much i can do about this from iptables , as i cannot restrict the source ip address for connection to the web ui, the best you can do is use a non common port number for the web ui thus reducing the chance of this happening.

you could also bind qbittorrent to the vpn adapter to force all traffic down the vpn, there are instruction on how to do this on the web.

**Comment by ttikaman** [comment]
Okay understood, thank you. From a privacy/security perspective do you think it would make sense for 8080 to be restricted to local network by default and have an additional flag allow public access? I would imagine many users who specifically selected this container for the VPN support would prefer maximum privacy by default.

**Comment by fryfrog** [comment]
This isn't great, do your other torrent client + vpn containers have this leak? Or is it unique to qB? Or just more common w/ qB since port 8080 is so frequently used for... everything? It seems like blocking it by default would be ideal, to keep the container secure. You could add a bypass for those that would prefer less secure, but working. :|

**Comment by binhex** [comment]
> do your other torrent client + vpn containers have this leak?

yes, they are all based off the same intermediate image.

> Or just more common w/ qB since port 8080 is so frequently used for... everything?

correct, there is a higher probability of this happening with qbittorrent, assuming you use the default port on the host side and of course assuming the tracker you connect to is using the same port and the VPN connection goes down whilst you have torrents in an active state.

> It seems like blocking it by default would be ideal, to keep the container secure.

after much deliberation and investigation, i have come to the same conclusion, however this will obviously increase the support levels for me dramatically when i make this change as everybody who updates will loose remote access to webui and also any other containers running through this one (or one of my other vpn containers)  - additional_ports are also susceptible to this.

so i have code to block both webui port and additional port(s) to lan and docker network only, i would like a bit of help testing this out to ensure it does fix the issue, so i probably will release a 'test' tagged release fist for testing before the release to the wider community.

i will of course need to replicate the same code changes to all of my vpn related images, and sadly this will mean quite a bit of work for me as iptables are defined per application so expect some time to get this in place for all apps.

please also keep in mind making something 100% secure whilst also being flexible to allow remote access is a VERY tricky game, especially when you put in the mix that people also want to route other container traffic through these vpn containers, this is akin to port forwarding on your home router - the more holes you punch the higher the risk of attack, its just how it is, nothing is 100% secure but i am def keen on making this as tight as possible, not only for you guys but for myself too.

**Comment by fryfrog** [comment]
Thanks for the details, I'll be happy to give it some testing when your `:tag` is ready. :)

**Comment by r0arkin** [comment]
I can help test as well.

**Comment by mrhotio** [comment]
If you just don't do this (https://github.com/binhex/arch-qbittorrentvpn/blob/0cdc1adc47a4de250c3abf0afc0fabc56c4fb8ff/run/root/iptable.sh#L212), it'll be fixed.

**Comment by binhex** [comment]
> If you just don't do this, it'll be fixed:

firstly, thanks for the post!, always nice to hear from a fellow docker dev, i think i have potentially been over engineering this and your simple solution i think would suffice, i will do some testing, thanks!.

**Comment by binhex** [comment]
a quick update on this, so whilst i am in there tweaking iptables.sh i decided to do a rework of the code so that i can incorporate it into the intermediate image so i don't have to deal with multiple copies, obviously this leads to additional testing which i have nearly completed, i hope to have a new test image for you guys by the end of today.

**Comment by binhex** [comment]
fighting with glibc and runc tonight ughh!, ok test image built:-
```
docker pull binhex/arch-qbittorrentvpn:test
```

please let me know how it goes, if it looks ok then i will tag as latest

**Comment by fryfrog** [comment]
Running. I'm not sure I'll notice any leaks because I run on a weird port (I hate software that uses 8080 and always change it), but the webui works via reverse proxy which is a good sign.

**Comment by binhex** [comment]
@fryfrog thanks for the quick feedback, i did do VERY extensive testing of the image, revsrse proxy. port forward, routing containers through it, the works!, so im reasonably confident it should work as intended, but its always nice to have a bit of feedback to confirm all is good.

**Comment by binhex** [comment]
ok last set of tweaks pushed up to test tagged image, this is a tightening of iptables for vpn endpoint communication. 

Unless i hear otherwise then i will be releasing the changes to tag 'latest' 2 days from now, PLEASE let me know if you see anything not working as intended, especially if it was previously working with the current 'latest' tagged image.

**Comment by fryfrog** [comment]
```
2021-02-23 08:41:06,274 DEBG 'start-script' stdout output:
[debug] OpenVPN command line:- /usr/bin/openvpn --reneg-sec 0 --mute-replay-warnings --auth-nocache --setenv VPN_PROV 'custom' --setenv VPN_CLIENT 'openvpn' --setenv DEBUG 'true' --setenv VPN_DEVICE_TYPE 'tun0' --setenv VPN_ENABLED 'yes' --setenv VPN_REMOTE_SERVER '206.189.218.112' --setenv APPLICATION 'qbittorrent' --script-security 2 --writepid /root/openvpn.pid --remap-usr1 SIGHUP --log-append /dev/stdout --pull-filter ignore 'up' --pull-filter ignore 'down' --pull-filter ignore 'route-ipv6' --pull-filter ignore 'ifconfig-ipv6' --pull-filter ignore 'tun-ipv6' --pull-filter ignore 'dhcp-option DNS6' --pull-filter ignore 'persist-tun' --pull-filter ignore 'reneg-sec' --up /root/openvpnup.sh --up-delay --up-restart --keepalive 10 60 --auth-user-pass credentials.conf --cd /config/openvpn --config '/config/openvpn/TorGuard.ovpn'
[info] Starting OpenVPN (non daemonised)...

2021-02-23 08:41:06,298 DEBG 'start-script' stdout output:
2021-02-23 08:41:06 WARNING: file 'credentials.conf' is group or others accessible
2021-02-23 08:41:06 OpenVPN 2.5.0 [git:makepkg/a73072d8f780e888+] x86_64-pc-linux-gnu [SSL (OpenSSL)] [LZO] [LZ4] [EPOLL] [PKCS11] [MH/PKTINFO] [AEAD] built on Nov  6 2020
2021-02-23 08:41:06 library versions: OpenSSL 1.1.1j  16 Feb 2021, LZO 2.10

2021-02-23 08:41:06,299 DEBG 'start-script' stdout output:
2021-02-23 08:41:06 NOTE: the current --script-security setting may allow this configuration to call user-defined scripts

2021-02-23 08:41:06,300 DEBG 'start-script' stdout output:
2021-02-23 08:41:06 TCP/UDP: Preserving recently used remote address: [AF_INET]206.189.218.112:1912
2021-02-23 08:41:06 UDP link local: (not bound)
2021-02-23 08:41:06 UDP link remote: [AF_INET]206.189.218.112:1912

2021-02-23 08:41:06,301 DEBG 'start-script' stdout output:
2021-02-23 08:41:06 write UDP: Operation not permitted (code=1)

2021-02-23 08:41:08,447 DEBG 'start-script' stdout output:
2021-02-23 08:41:08 write UDP: Operation not permitted (code=1)

2021-02-23 08:41:12,739 DEBG 'start-script' stdout output:
2021-02-23 08:41:12 write UDP: Operation not permitted (code=1)
````

^ This one isn't working for me, I'm starting to dig into it now.

**Comment by binhex** [comment]
hmm ok the write UDP: Operation not permitted (code=1) is normally a symptom of iptables blocking the connection to the vpn endpoint.

ok can you do the following:- https://github.com/binhex/documentation/blob/master/docker/faq/help.md

**Comment by fryfrog** [comment]
https://paste.ubuntu.com/p/mwPnwthSz3/

^ Hopefully I redacted all the good stuff.

**Comment by binhex** [comment]
replicated the issue and fixed, i will let you know once the image is ready.

**Comment by binhex** [comment]
ok it's there, please pull down 'test' tagged again.

**Comment by fryfrog** [comment]
This one is doing good, webui works, its uploading and sonarr/radarr are happily connected. I'm still not able to test for the same leaks as OP and I need to find something to download, but I think we know that's gonna be fine.

**Comment by binhex** [comment]
ok great!, well like i mentioned i will leave this till thursday and then press the button to build latest if no issues, mayybe @ttikaman or @r0arkin can check in the meantime.

**Comment by r0arkin** [comment]
I'll test this out shortly!

**Comment by r0arkin** [comment]
Okay, I pulled the test image and everything seems to work great. I use Mullvad with Wireguard and no issues seen. Everything looks connectable. Also webui works via reverse proxy and the custom port I set it to.

**Comment by binhex** [comment]
now released as tagged 'latest' im closing this issue off, any other side effects please raise a new issue.
